pkg/scheduler/queue/queue.go

```diff
@@ -54,229 +54,354 @@ type Request interface{}
-}
-
-func (q *RequestQueue) UnregisterQuerierConnection(querier string) {
-	q.connectedQuerierWorkers.Dec()
-
-	q.mtx.Lock()
-	defer q.mtx.Unlock()
-	q.queues.removeQuerierConnection(querier, time.Now())
-}
+	op := querierOperation{
+		querierID: querierID,
+		operation: registerConnection,
+		processed: make(chan struct{}),
+	}
 
-func (q *RequestQueue) NotifyQuerierShutdown(querierID string) {
-	q.mtx.Lock()
-	defer q.mtx.Unlock()
-	q.queues.notifyQuerierShutdown(querierID)
+	select {
+	case q.querierOperations <- op:
+		<-op.processed
+	case <-q.stopCompleted:
+		// TODO: return error?
+	}
 }
 
-func (q *RequestQueue) GetConnectedQuerierWorkersMetric() float64 {
-	return float64(q.connectedQuerierWorkers.Load())
-}
+func (q *RequestQueue) UnregisterQuerierConnection(querierID string) {
+	q.connectedQuerierWorkers.Inc()
 
-// contextCond is a *sync.Cond with Wait() method overridden to support context-based waiting.
-type contextCond struct {
-	*sync.Cond
+	op := querierOperation{
+		querierID: querierID,
+		operation: unregisterConnection,
+		processed: make(chan struct{}),
+	}
 
-	// testHookBeforeWaiting is called before calling Cond.Wait() if it's not nil.
-	// Yes, it's ugly, but the http package settled jurisprudence:
-	// https://github.com/golang/go/blob/6178d25fc0b28724b1b5aec2b1b74fc06d9294c7/src/net/http/client.go#L596-L601
-	testHookBeforeWaiting func()
+	select {
+	case q.querierOperations <- op:
+		<-op.processed
+	case <-q.stopCompleted:
+		// TODO: return error?
+	}
 }
 
-// Wait does c.cond.Wait() but will also return if the context provided is done.
-// All the documentation of sync.Cond.Wait() applies, but it's especially important to remember that the mutex of
-// the cond should be held while Wait() is called (and mutex will be held once it returns)
-func (c contextCond) Wait(ctx context.Context) {
-	// "condWait" goroutine does q.cond.Wait() and signals through condWait channel.
-	condWait := make(chan struct{})
-	go func() {
-		if c.testHookBeforeWaiting != nil {
-			c.testHookBeforeWaiting()
-		}
-		c.Cond.Wait()
-		close(condWait)
-	}()
-
-	// "waiting" goroutine: signals that the condWait goroutine has started waiting.
-	// Notice that a closed waiting channel implies that the goroutine above has started waiting
-	// (because it has unlocked the mutex), but the other way is not true:
-	// - condWait it may have unlocked and is waiting, but someone else locked the mutex faster than us:
-	//   in this case that caller will eventually unlock, and we'll be able to enter here.
-	// - condWait called Wait(), unlocked, received a broadcast and locked again faster than we were able to lock here:
-	//   in this case condWait channel will be closed, and this goroutine will be waiting until we unlock.
-	waiting := make(chan struct{})
-	go func() {
-		c.L.Lock()
-		close(waiting)
-		c.L.Unlock()
-	}()
+func (q *RequestQueue) NotifyQuerierShutdown(querierID string) {
+	q.connectedQuerierWorkers.Inc()
```

Comment:
should the counter move here? It wasn't moving in the previous implementation

Reply:
Yep, good catch, this was a copy-pasta.